numpy/core/src/common/npstd.hpp

```diff
@@ -34,23 +36,439 @@ using std::complex;
+                             ID == kULong || ID == kULongLong;
+
+/// Checks whether `T` is an signed integer type.
+template<typename T, TypeID ID = kTypeID<T>>
+constexpr bool kIsSigned = ID == kByte || ID == kShort ||
+                           ID == kLong || ID == kLongLong;
+
+/// Checks whether `T` is an integer type.
+template<typename T, TypeID ID = kTypeID<T>>
+constexpr bool kIsIntegral = kIsSigned<T> || kIsUnsigned<T>;
+
+/// Checks whether `T` is an floating-point type.
+template<typename T, TypeID ID = kTypeID<T>>
+constexpr bool kIsFloat = ID == kHalf || ID == kFloat ||
+                          ID == kDouble || ID == kLongDouble;
+
+/// Checks whether `T` is an complex type.
+template<typename T, TypeID ID = kTypeID<T>>
+constexpr bool kIsComplex = ID == kCFloat || ID == kCDouble ||
+                            ID == kCLongDouble;
+
+/// Checks whether `T` is an time type (`DateTime`, `Timedelta`).
+template<typename T, TypeID ID = kTypeID<T>>
+constexpr bool kIsTime = ID == kDateTime || ID == kTimeDelta;
+
+/// @} cpp_core_constants
+
+
+/// @addtogroup cpp_core_containers
+/// @{
+/// Class wraps builtin data types.
+/// To distinct them, and gives the room for dealing
+/// with platform compatiblty.
+template <typename T, TypeID>
+class Holder {
+  public:
+    /// initlize nothing.
+    Holder() = default;
+    /// Construct from `T`
+    constexpr Holder(const T &val) : val_(val)
+    {}
+    /// cast
+    constexpr operator T () const
+    { return val_; }
+
+  private:
+    T val_;
+};
+template<template<class> class C, typename T, TypeID ID>
+class Holder<C<T>, ID> : public C<T> {
+  public:
+    using C<T>::C;
+};
+/// @} cpp_core_containers
+
+/// @addtogroup cpp_core_types
+/// @{
+/// Signed integer with at least 8-bit width.
+using Byte = Holder<signed char, kByte>;
+/// Unsigned integer with at least 8-bit width.
+using UByte = Holder<unsigned char, kUByte>;
+/// Signed integer with at least 16-bit width.
+using Short = Holder<short, kShort>;
+/// Unsigned integer with at least 16-bit width.
+using UShort = Holder<unsigned short, kUShort>;
+/// Signed integer with at least 16-bit width.
+using Int = Holder<int, kInt>;
+/// Unsigned integer with at least 16-bit width.
+using UInt = Holder<unsigned int, kUInt>;
+/// Signed integer with at least 32-bit width.
+using Long = Holder<long, kLong>;
+/// Unsigned integer with at least 32-bit width.
+using ULong = Holder<unsigned long, kULong>;
+/// Signed integer with at least 64-bit width.
+using LongLong = Holder<long long, kLongLong>;
+/// Unsigned integer with at least 64-bit width.
+using ULongLong = Holder<unsigned long long, kULongLong>;
+/// Boolean type, stored as `Byte`.
+/// It may only be set to the values false and true.
+using Bool = std::conditional_t<sizeof(bool) == sizeof(unsigned char),
+    bool, Holder<unsigned char, kBool>>;
+/// single-precision floating-point type, compatible with IEEE-754.
+using Float = float;
```

Comment:
Why is `Holder` used for `long` but not `float`?

Reply:
Theoretically, only the following data types are required to be distinguished:
```C++
using UByte = Holder<unsigned char, kUByte>;

using Bool = std::conditional_t<sizeof(bool) == sizeof(unsigned char),
    bool, Holder<unsigned char, kBool>>;

using LongDouble = std::conditional_t<kLongDoubleIsDouble,
      Holder<double, kLongDouble>, long double>;

using CLongDouble = std::conditional_t<kLongDoubleIsDouble,
      Holder<complex<double>, kCLongDouble>, complex<long double>>;
```


To forbid the use of built-in datatypes directly, using `long` instead of `Long` should break the build when `kTypeID` is involved:

https://github.com/numpy/numpy/blob/2beb25fa4e90923e7b260408bf45b81fe238cf44/numpy/core/src/common/npstd.hpp#L174C4-L176


Allowing raw integer built-in data types isn't type-safe when it comes to determining the data type IDs (template specialization), since they are commonly used as alias almost for all the generic std alais including fixed width data types.